### ISSUE TITLE
Error when unzipping: "A required privilege is not held by the client"

### ISSUE THREAD

**Issue by ogerardin** [maven-download-plugin/maven-download-plugin#165]
Full error: 

`[ERROR] Failed to execute goal com.googlecode.maven-download-plugin:download-maven-plugin:1.6.0:wget (download-jar2app) on project xpman-fx-dist: IO Error: Error while expanding xxxxx\target\jar2app\master.zip: xxxxx\target\jar2app\jar2app-master\jar2app: A required privilege is not held by the client. -> [Help 1]
`

The ZIP can be unzipped without error by 7zip or even Windows builtin unzipper.

Plugin configuration:

```` 
                      <plugin>
                        <groupId>com.googlecode.maven-download-plugin</groupId>
                        <artifactId>download-maven-plugin</artifactId>
                        <version>1.6.0</version>
                        <executions>
                            <!-- download and unpack jar2app -->
                            <execution>
                                <id>download-jar2app</id>
                                <phase>prepare-package</phase>
                                <goals>
                                    <goal>wget</goal>
                                </goals>
                                <configuration>
                                    <url>https://github.com/Jorl17/jar2app/archive/master.zip</url>
                                    <outputDirectory>${project.build.directory}/jar2app</outputDirectory>
                                    <unpack>true</unpack>
                                </configuration>
                            </execution>
                        </executions>
                    </plugin>
````

Maven config:

```
Apache Maven 3.6.3 (cecedd343002696d0abb50b32b541b8a6ba2883f)
Maven home: C:\Users\Olivier\scoop\apps\maven\current\bin\..
Java version: 1.8.0_251, vendor: Oracle Corporation, runtime: C:\Program Files\Java\jdk1.8.0_251\jre
Default locale: en_GB, platform encoding: Cp1252
OS name: "windows 10", version: "10.0", arch: "amd64", family: "windows"
```

**Comment by longtimeago** [comment]
@ogerardin does this happen with any zip (with any content inside)? or just with this particular zip?

**Comment by ogerardin** [comment]
> @ogerardin does this happen with any zip (with any content inside)? or just with this particular zip?

I haven't tested with another ZIP recently but I've used it successfully in the past. Can you reproduce it with my config at least?

**Comment by longtimeago** [comment]
@ogerardin unfortunately, I can only guess because I have no win PC :(

**Comment by ogerardin** [comment]
> @ogerardin unfortunately, I can only guess because I have no win PC :(

You could use a VM...

**Comment by longtimeago** [comment]
@ogerardin that would take a time, which I don't have now unfortunately (

**Comment by joachim-durchholz-six** [comment]
For me, the message happens on the first symbolic link inside a .tar.gz file.

**Comment by joachim-durchholz-six** [comment]
Relevant messages from Process Monitor.

Baseline is the successful creation of the `C:\develop\projects\id-gui\target\downloads\zulu11.41.24-sa-jdk11.0.8-linux_x64\man` directory:

```
QueryDirectory	C:\develop\projects\id-gui\target\downloads\zulu11.41.24-sa-jdk11.0.8-linux_x64\man	SUCCESS	Filter: man, 1: man, FileInformationClass: FileBothDirectoryInformation
CloseFile	C:\develop\projects\id-gui\target\downloads\zulu11.41.24-sa-jdk11.0.8-linux_x64	SUCCESS	
CreateFile	C:\develop\projects\id-gui\target\downloads\zulu11.41.24-sa-jdk11.0.8-linux_x64\man	SUCCESS	Desired Access: Read Data/List Directory, Synchronize, Disposition: Open, Options: Directory, Synchronous IO Non-Alert, Attributes: n/a, ShareMode: Read, Write, Delete, AllocationSize: n/a, OpenResult: Opened
```

Failed creation of the `C:\develop\projects\id-gui\target\downloads\zulu11.41.24-sa-jdk11.0.8-linux_x64\man\ja` directory symlink:

```
QueryDirectory	C:\develop\projects\id-gui\target\downloads\zulu11.41.24-sa-jdk11.0.8-linux_x64\man\ja	NO SUCH FILE	Filter: ja
CloseFile	C:\develop\projects\id-gui\target\downloads\zulu11.41.24-sa-jdk11.0.8-linux_x64\man	SUCCESS	
CreateFile	C:\develop\projects\id-gui\target\downloads\zulu11.41.24-sa-jdk11.0.8-linux_x64\man\ja	NAME NOT FOUND	Desired Access: Read Attributes, Disposition: Open, Options: Open Reparse Point, Attributes: n/a, ShareMode: Read, Write, Delete, AllocationSize: n/a
```

**Comment by joachim-durchholz-six** [comment]
Cygwin tar with  `CYGWIN=winsymlinks:nativestrict` (forcing Cygwin to create symlinks using Windows mechanisms instead of a workaround) gives error messages, too. Here's a sample with first the directory, then a file:

```
tar: zulu11.41.24-sa-jdk11.0.8-linux_x64/man/ja: Cannot create symlink to ‘ja_JP.UTF-8’: No such file or directory
tar: zulu11.41.24-sa-jdk11.0.8-linux_x64/legal/java.smartcardio/LICENSE: Cannot create symlink to ‘../java.base/LICENSE’: Operation not permitted
```

**Comment by joachim-durchholz-six** [comment]
Winzip unpacks all symlinks (directory or not) as zero-length files, so it does not demonstrate how to do it :-(

**Comment by joachim-durchholz-six** [comment]
It is indeed a permission problem: On the Windows command line, I get this:

```
C:\develop\projects\id-gui\target\downloads>mklink test zulu11.41.24-sa-jdk11.0.8-linux_x64
You do not have sufficient privilege to perform this operation.
```

(`mklink` is described in https://www.howtogeek.com/howto/16226/complete-guide-to-symbolic-links-symlinks-on-windows-or-linux/; basic command syntax is `mklink Link Target`.)

This concludes my analysis: It is indeed a permission problem, Windows requires administrator rights to create symlinks.
(The linked article states that the "Windows 10 Creator's Update" has this restriction lifted. I can't test because I am on a managed machine where I cannot have administrator rights.)

**Comment by joachim-durchholz-six** [comment]
Can we have an unpack option that merely duplicates symlinked content, instead of failing with an error?
It does not serve all use cases so it needs to be an opt-in option, but at least it's there for the situations where the files just need to be there.

**Comment by joachim-durchholz-six** [comment]
Ah. I see Plexus UnArchiver does not support expanding symlinks, and the affected code would need some restructuring (first off, it would have to keep track of all symlinks encountered, and copy the file contents over as a final pass, because otherwise the symlinked files may not have been unpacked yet).

I'm not going to pursue this further, so feel free to push for such a change in Plexus, or decline the enhancement request given that it's a permissions problem.
(In the latter case, a slight improvement might be to improve the error message so Windows users know to consider disabled symlink creation as a potential cause.)

If you got this far: Thanks for reading!
And sorry for my inability to find a straightforward solution.

**Comment by longtimeago** [comment]
@joachim-durchholz-six thank you very much for the investigation!
In the mentioned article I saw an option "However, to do this without an Administrator Command Prompt window, you must first enable Developer Mode from Settings > Update & Security > For Developers."
Wonder if that could help

Otherwise, even patching plexus will not help IMO

**Comment by kriegaex** [comment]
I think, this issue should first be renamed to reflect the fact that the problem is about unpacking archives with symlinks on Windows. Secondly, the issue seems to be identified, and it is clear that the problem is not in the plugin itself. Therefore, the issue should be closed.

**Comment by joachim-durchholz-six** [comment]
@longtimeago
The plexus change was supposed to give it the ability to unpack symlinks as files.
This would allow the download plugin to offer that approach as an option, for those use cases where such a fallback is acceptable.

@kriegaex The permission problem is not in the plugin, but the plugin does a poor job at informing the user what's going on, and nothing at all to help him deal with the situation.
The decision to close an issue should not just consider the root problem, but also what mitigations are possible.

**Comment by ogerardin** [comment]
> I think, this issue should first be renamed to reflect the fact that the problem is about unpacking archives with symlinks on Windows. 
Agreed, I changed the title.

> Secondly, the issue seems to be identified, and it is clear that the problem is not in the plugin itself. Therefore, the issue should be closed.
As @joachim-durchholz-six said, at least the error message should be more informative, but i think the plugin should be configurable to issue warnings instead of failing, or to expand symlinks as a copy of the original file if it's in the archive.

**Comment by kriegaex** [comment]
Isn't that level of sophistication somewhat out of scope for a **download** plugin in which unpacking is just a basic add-on option? I know, I said something similar [here](https://github.com/maven-download-plugin/maven-download-plugin/issues/151#issuecomment-833292427) already.

**Comment by ogerardin** [comment]
> Isn't that level of sophistication somewhat out of scope for a **download** plugin in which unpacking is just a basic add-on option? I know, I said something similar [here](https://github.com/maven-download-plugin/maven-download-plugin/issues/151#issuecomment-833292427) already.

You're right that downloading and unpacking are different concerns. However, either you do something and then you do it reliably, or you don't. Since the original problematic zip can be extracted by the very basic Windows extractor, I expect that this plugin should be able to do the same.

**Comment by kriegaex** [comment]
Fair enough, point taken. As a workaround - either temporary until this is fixed or permanent if it does not get fixed - you can use TrueZIP Maven Plugin. It unpacks the symlink like e.g. the Windows 10 on-board unpacker and 7-zip do it, too: The symlink `jar2app` in your sample archive becomes a small text file with the content `jar2app.py`, i.e. the name of the target the symlink points to. BTW, only the `unzip` UNIX command line utility which is installed on Windows with Git Bash duplicates the file content. Symlinks pointing to directories will be unpacked the same way, i.e. files with link target as content, no duplication of directory contents.
